mixer/pkg/adapter/test/integration.go

```diff
@@ -0,0 +1,270 @@
+package test
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"log"
+	"path"
+	"path/filepath"
+	"reflect"
+	"runtime"
+	"testing"
+
+	"google.golang.org/grpc"
+
+	istio_mixer_v1 "istio.io/api/mixer/v1"
+	"istio.io/istio/mixer/pkg/adapter"
+	"istio.io/istio/mixer/pkg/attribute"
+	"istio.io/istio/mixer/pkg/config/storetest"
+	"istio.io/istio/mixer/pkg/server"
+	"istio.io/istio/mixer/pkg/template"
+)
+
+type (
+	// TestCase fully defines an adapter integration test
+	TestCase struct {
+		// Cfgs is a list of CRDs that Mixer will read.
+		Cfgs []string
+		// ParallelCalls is a list of test calls to be made to Mixer
+		// in parallel.
+		ParallelCalls []Call
+		// Expected json string for a WantObject object.
+		//
+		// New test can start of with an empty "{}" string and then
+		// get the baseline from the failure logs upon execution.
+		Want string
+	}
+	// Call represents the input to make a call to Mixer
+	Call struct {
+		// CallKind can be either CHECK or REPORT
+		CallKind CallKind
+		// Attrs to call the Mixer with.
+		Attrs map[string]interface{}
+		// Quotas info to call the Mixer with.
+		Quotas map[string]istio_mixer_v1.CheckRequest_QuotaParams
+	}
+	// CallKind represents the call to make; check or report.
+	CallKind int32
+
+	// Result represents the test baseline
+	Result struct {
+		// AdapterState represents adapter specific baseline data
+		AdapterState interface{} `json:"AdapterState"`
+		// Returns represents the return data from calls to Mixer
+		Returns []Return `json:"Returns"`
+	}
+	// Return represents the return data from a call to Mixer
+	Return struct {
+		// Check is the response from a check call to Mixer
+		Check adapter.CheckResult `json:"Check"`
+		// Quota is the response from a check call to Mixer
+		Quota map[string]adapter.QuotaResult `json:"Quota"`
+		// Error is the error from call to Mixer
+		Error error `json:"Error"`
+	}
+)
+
+const (
+	// CHECK for  Mixer Check
+	CHECK CallKind = 0
+	// REPORT for  Mixer Report
+	REPORT CallKind = 1
+)
+
+type (
+	setupFn    func() (ctx interface{}, err error)
+	teardownFn func(ctx interface{})
+	getStateFn func(ctx interface{}) (interface{}, error)
+)
```

Comment:
Should these be capital, because they are types for parameters on an exported function?